primitives/io/src/lib.rs

```diff
@@ -109,8 +109,8 @@ pub enum KillStorageResult {
 #[runtime_interface]
 pub trait Storage {
 	/// Returns the data for `key` in the storage or `None` if the key can not be found.
-	fn get(&self, key: &[u8]) -> Option<Vec<u8>> {
-		self.storage(key).map(|s| s.to_vec())
+	fn get(&self, key: &[u8]) -> Option<bytes::Bytes> {
+		self.storage(key).map(|s| bytes::Bytes::from(s.to_vec()))
```

Comment:
That's the whole problem that this kind of an optimization *wasn't* used. (:

Basically, we have `Vec<u8>` special-cased so that it doesn't need to go through SCALE but we've missed the `Option<Vec<u8>>` which *does* need to go through SCALE. We can't special-case it now since it'll break existing chains, so this PR does the next best thing and replaces the `Option<Vec<u8>>` with `Option<Bytes>`. Both go through SCALE and have exactly the same SCALE representation however deserializing `Option<Bytes>` doesn't need to unnecessarily allocate the payload a second time.

Reply:
I did suspect it a bit, thanks for confirming. Certainly worth changing it at a latter point if new host function are created (there is the new storage_kill one that was recently changed and probably not already used, but not sure if worth changing it at this point (it is only an input parameter of a function that should not be call frequently))